x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecyclePolicyUtils.java

```diff
@@ -109,26 +112,36 @@ public static ItemUsage calculateUsage(
             .map(indexMetadata -> indexMetadata.getIndex().getName())
             .collect(Collectors.toList());
 
+        // First find all the index templates that use this policy, and sort them descending on priority.
+        final var composableTemplates = state.metadata().templatesV2().entrySet().stream().filter(entry -> {
+            Settings settings = MetadataIndexTemplateService.resolveSettings(entry.getValue(), state.metadata().componentTemplates());
+            return policyName.equals(LifecycleSettings.LIFECYCLE_NAME_SETTING.get(settings));
+        }).sorted(Comparator.comparing(entry -> entry.getValue().priorityOrZero(), Comparator.reverseOrder())).toList();
+
+        // These index templates are returned as a type of usage themselves.
+        final var composableTemplateNames = composableTemplates.stream().map(Map.Entry::getKey).toList();
+
         final List<String> allDataStreams = indexNameExpressionResolver.dataStreamNames(
             state,
             IndicesOptions.LENIENT_EXPAND_OPEN_CLOSED_HIDDEN
         );
 
+        // We filter all the data streams by finding the first index template (highest priority) whose index pattern covers the data stream.
         final List<String> dataStreams = allDataStreams.stream().filter(dsName -> {
-            String indexTemplate = MetadataIndexTemplateService.findV2Template(state.metadata(), dsName, false);
```

Comment:
I considered refactoring `MetadataIndexTemplateService#findV2CandidateTemplates` instead to make it take a list of templates as a parameter (rather than always looping over all templates in the cluster state). That would allow for a speedup [here](https://github.com/elastic/elasticsearch/blob/d160dc2255586953299c256b744c319fa8a6a349/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java#L1149-L1149) as well, but I couldn't come up with a way to also speed up `MetadataIndexTemplateService#findV2Template` at the same time, so I decided to do it this way. Let me know if others think it's worth refactoring that method anyway and using it here as well.